BlueBrain/Search
# Back up MySQL database

Fixes #354 .

# Description
As commented in #354, a sensible approach is described [here](https://jpaulmora.github.io/2020/03/22/how-to-properly-backup-mysql/):
- Cron job dumps whole database to binary files using [`mysqldump`](https://dev.mysql.com/doc/mysql-backup-excerpt/8.0/en/using-mysqldump.html) daily/weekly.
- MySQL [binary log](https://dev.mysql.com/doc/refman/8.0/en/binary-log.html) records incremental changes to DB between file backups.

### In this PR:
- We add the script `make-backup` to the `mysql` docker image that can be run to create database dumps
- This script should be called via docker: `docker exec -t mysql-container make-backup MYSQL_USER MYSQL_PASSWORD DB_NAME`.
- The backup files will be placed into the `/backup` directory within the container. This directory is configured to be an anonymous local docker volume, but it's better to explicitly mount a volume into that directory via the `-v` parameter in `docker run`.
- We configure the binary logs to have an expiry of 7 days

### Not in this PR: cron job
We should try and configure an external cron job that calls the `docker exec` command above every 7 days if we want to make backups that way.

I experimented with having a cron service running within the mysql container, but
- it's not really recommended that a docker container does more than one thing / runs more than one process
- one would need to store the MySQL password on the container in order to be able to call `mysqldump`
- one would need to know upfront the names of the databases that will be backed up (or back up all databases)

If one is happy with the above drawbacks, then we can still go for that solution. Actually, the first commit in this PR (2b354da) already did all the necessary preparations apart from adding the actual cron job (e.g. just a script in `/etc/cron.weekly`):
- Installing the cron and syslog services
- Starting the cron and syslog services at container startup

# How to test?
Build a test image
```bash
docker build -f docker/mysql.Dockerfile -t my-test-mysql-image .
```

Start a container
```bash
mkdir mysql-backup
docker run \
--rm \
-d \
-e MYSQL_ROOT_PASSWORD=whatever \
-e MYSQL_DATABASE=testdb \
--name my-test-mysql-server \
-v $PWD/mysql-backup:/backup \
-P \
my-test-mysql-image
```
(normally one would probably also mount the `/var/lib/mysql` directory, but it's not necessary here)

Check if it's running via `docker ps`. If not try replacing `-d` by `-it` and checking what's wrong.

Populate the database with some random data:
```python
import random
import string

import sqlalchemy

engine = sqlalchemy.create_engine("mysql://root:whatever@<host>:<port>/testdb")
with engine.begin() as conn:
    conn.execute("drop table if exists mytable; create table mytable(name text, age int);")
    for _ in range(100):
        name = "".join(random.sample(string.ascii_letters, 10))
        age = random.randint(18, 90)
        conn.execute(f"insert into mytable(name, age) values ('{name}', {age})")
```

### Test the file backup
Trigger a backup 
```
export MYSQL_USER=root
export MYSQL_PWD=whatever
docker exec -t -e MYSQL_USER -e MYSQL_PWD my-test-mysql-server make-backup testdb
```

Check the backup is there:
```
ls mysql-backup
```

### Test recovery from backup
Drop the database
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'drop database testdb'
```

Check database doesn't exist
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'select count(*) from testdb.mytable'
Enter password: 
ERROR 1049 (42000) at line 1: Unknown database 'testdb'
```

Recover form backup file
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'create database testdb'
docker exec -i my-test-mysql-server mysql -uroot -p$MY_PWD testdb < /path/to/backup.sql
```

Check data is there
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'select count(*) from testdb.mytable'
Enter password: 
+----------+
| count(*) |
+----------+
|      100 |
+----------+
```

### Test recovery from binlog files
More info here: https://dev.mysql.com/doc/refman/8.0/en/point-in-time-recovery.html

Check binlog is enabled
```
docker exec -it my-test-mysql-server mysql -uroot -p -e "show variables like 'log_bin'"
Enter password: 
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| log_bin       | ON    |
+---------------+-------+
```

Check data is there
```
docker exec -it my-test-mysql-server mysql -uroot -p -e 'select count(*) from testdb.mytable'
Enter password: 
+----------+
| count(*) |
+----------+
|      100 |
+----------+
```

Get current position in the binlog
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'show master status'
Enter password: 
+---------------+----------+--------------+------------------+-------------------+
| File          | Position | Binlog_Do_DB | Binlog_Ignore_DB | Executed_Gtid_Set |
+---------------+----------+--------------+------------------+-------------------+
| binlog.000002 |    15473 |              |                  |                   |
+---------------+----------+--------------+------------------+-------------------+
```

Drop the database
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'drop database testdb'
```

Check database doesn't exist
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'select count(*) from testdb.mytable'
Enter password: 
ERROR 1049 (42000) at line 1: Unknown database 'testdb'
```

Replay all binlog records up to the point where we dropped the database
```bash
docker exec -it my-test-mysql-server bash
cd /var/lib/mysql/
mysqlbinlog --stop-position 15473 binlog.000001 binlog.000002 | mysql -uroot -p
exit
```

Check data is there
```
docker exec -it my-test-mysql-server mysql -uroot -p -e 'select count(*) from testdb.mytable'
Enter password: 
+----------+
| count(*) |
+----------+
|      100 |
+----------+
```

# Checklist

  (if it is not the case, please create an issue first).
  (if needed)
  (if needed)
  (if needed)
  (if a function is added or modified)
